// @yuzu/decoratify/lib/TransformWriter.js
"use strict";

var path = require("path");
var ES_EXPORT_EXPR = "exports[\"default\"]";
var NODE_EXPORT_EXPR = "module.exports";

function TransformWriter(file, body) {
    var exportExpressions = [];
    var exportExpr = /module.exports/.test(body) ? NODE_EXPORT_EXPR : ES_EXPORT_EXPR;

    function addExportExpression(expression, operator) {
        operator = operator || ".";
        exportExpressions.push(exportExpr + operator + expression + ";");
    }

    this.decorateExport = function (type, args) {
        var expr = "(" + type + "(" + args.join(", ") + ")(" + exportExpr + ")) || " + exportExpr;
        addExportExpression(expr, " = ");
    };

    this.appendToDefaultExport = addExportExpression;

    this.appendToBody = function (content) {
        body += content + "\n";
    };

    Object.defineProperties(this, {
        file: { value: file },
        filename: { value: path.basename(file) },
        body: {
            get: function () {
                var footer = exportExpressions.length > 0 ? exportExpressions.join("\n") + "\n" : "";
                return body + footer;
            }
        }
    });
}

TransformWriter.ES_EXPORT_EXPR = ES_EXPORT_EXPR;
TransformWriter.NODE_EXPORT_EXPR = NODE_EXPORT_EXPR;

module.exports = TransformWriter;
