// workflowit/lib/Workflow.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var _q = require("q");

var _q2 = _interopRequireDefault(_q);

var _pubitAsPromised = require("pubit-as-promised");

var Step = (function () {
    function Step(name, fn) {
        _classCallCheck(this, Step);

        this.name = name;
        this.fn = fn;
        this._publish = (0, _pubitAsPromised.makeEmitter)(this, ["exec"]);
    }

    _createClass(Step, [{
        key: "exec",
        value: function exec(wf) {
            var _this = this;

            return (0, _q2["default"])(this.fn(this, wf)).then(function () {
                return _this._publish("exec", wf);
            });
        }
    }]);

    return Step;
})();

var Workflow = (function () {
    function Workflow(name) {
        _classCallCheck(this, Workflow);

        this._name = name;

        for (var _len = arguments.length, steps = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
            steps[_key - 1] = arguments[_key];
        }

        this._steps = steps || [];
        this._publish = (0, _pubitAsPromised.makeEmitter)(this, ["step", "run"]);
    }

    _createClass(Workflow, [{
        key: "addStep",
        value: function addStep(step) {
            this._steps.push(step);
            return this;
        }
    }, {
        key: "run",
        value: function run() {
            var _this2 = this;

            var chainSteps = function chainSteps(prev, nextStep) {
                return prev.then(function () {
                    return _this2._onStep(prev, nextStep);
                });
            };

            return this._steps.reduce(chainSteps, _q2["default"].resolve()).then(function () {
                return _this2._publish.when("run");
            });
        }
    }, {
        key: "_onStep",
        value: function _onStep(prev, step) {
            var _this3 = this;

            var progress = this._progressOf(step);
            return prev.then(function () {
                return step.exec(_this3);
            }).then(function () {
                return _this3._publish.when("step", step, progress);
            });
        }
    }, {
        key: "_progressOf",
        value: function _progressOf(step) {
            return (this._steps.indexOf(step) + 1) / this._steps.length;
        }
    }, {
        key: "name",
        get: function get() {
            return this._name;
        }
    }], [{
        key: "Step",
        get: function get() {
            return Step;
        }
    }]);

    return Workflow;
})();

exports["default"] = Workflow;
module.exports = exports["default"];
